pkg/helper/product_entity.go

```diff
@@ -0,0 +1,490 @@
+	return nil
+}
+
+func (b *ProductEntity) UpdateBackendUsage(id int64, params threescaleapi.Params) error {
+	b.logger.V(1).Info("UpdateBackendUsage", "ID", id, "params", params)
+	_, err := b.client.UpdateBackendapiUsage(b.productObj.Element.ID, id, params)
+	if err != nil {
+		return fmt.Errorf("product [%s] update backendusage: %w", b.productObj.Element.SystemName, err)
+	}
+	b.resetBackendUsages()
+	return nil
+}
+
+func (b *ProductEntity) CreateBackendUsage(params threescaleapi.Params) error {
+	b.logger.V(1).Info("CreateBackendUsage", "params", params)
+	_, err := b.client.CreateBackendapiUsage(b.productObj.Element.ID, params)
+	if err != nil {
+		return fmt.Errorf("product [%s] update backendusage: %w", b.productObj.Element.SystemName, err)
+	}
+	b.resetBackendUsages()
+	return nil
+}
+
+func (b *ProductEntity) Proxy() (*threescaleapi.ProxyJSON, error) {
+	b.logger.V(1).Info("Proxy")
+	if b.proxy == nil {
+		proxy, err := b.getProxy()
+		if err != nil {
+			return nil, err
+		}
+		b.proxy = proxy
+	}
+	return b.proxy, nil
+}
+
+func (b *ProductEntity) UpdateProxy(params threescaleapi.Params) error {
+	b.logger.V(1).Info("UpdateProxy", "params", params)
+	updated, err := b.client.UpdateProductProxy(b.productObj.Element.ID, params)
+	if err != nil {
+		return fmt.Errorf("product [%s] update proxy: %w", b.productObj.Element.SystemName, err)
+	}
+
+	b.proxy = updated
+	return nil
+}
+
+func (b *ProductEntity) ApplicationPlans() (*threescaleapi.ApplicationPlanJSONList, error) {
+	b.logger.V(1).Info("ApplicationPlans")
+	if b.plans == nil {
+		plans, err := b.getApplicationPlans()
+		if err != nil {
+			return nil, err
+		}
+		b.plans = plans
+	}
+	return b.plans, nil
+}
+
+func (b *ProductEntity) DeleteApplicationPlan(id int64) error {
+	b.logger.V(1).Info("DeleteApplicationPlan", "ID", id)
+	err := b.client.DeleteApplicationPlan(b.productObj.Element.ID, id)
+	if err != nil {
+		return fmt.Errorf("product [%s] delete applicationPlan: %w", b.productObj.Element.SystemName, err)
+	}
+	b.resetApplicationPlans()
+	return nil
+}
+
+func (b *ProductEntity) CreateApplicationPlan(params threescaleapi.Params) (*threescaleapi.ApplicationPlan, error) {
+	b.logger.V(1).Info("CreateApplicationPlan", "params", params)
+	obj, err := b.client.CreateApplicationPlan(b.productObj.Element.ID, params)
+	if err != nil {
+		return nil, fmt.Errorf("product [%s] create plan: %w", b.productObj.Element.SystemName, err)
+	}
+	b.resetApplicationPlans()
+	return obj, nil
+}
+
+func (b *ProductEntity) PromoteProxyToStaging() error {
+	b.logger.V(1).Info("PromoteProxyToStaging")
+	proxyObj, err := b.client.DeployProductProxy(b.productObj.Element.ID)
+	if err != nil {
+		return fmt.Errorf("product [%s] promote proxy to staging: %w", b.productObj.Element.SystemName, err)
+	}
+
+	b.proxy = proxyObj
+	return nil
+}
+
+//
+// PRIVATE
+//
+//
+
+func (b *ProductEntity) resetBackendUsages() {
+	b.backendUsages = nil
+}
+
+func (b *ProductEntity) resetMethods() {
```

Comment:
Why are all this reset* methods implemented in ProductEntity and BackendAPIEntity? Is it a caching mechanism that we invalidate when performing update/delete/create ? If so, why do we reset when creating or updating instead of updating the values?

Reply:
ProductEntity and BackendEntity implement some 3scale API related code and implement caching as well. 

Regarding caching, the http response body is cached that is basically array of objects. When you delete or create an object, you could manipulate (add, delete, update elements) the cached array to be more efficient. For simplicity, it is reseted. Can be a improvement, indeed.

Reply:
:+1: